### PR TITLE
feat: Enhance Observability with ArizePhoenixTracer Integration

### PR DESCRIPTION
**Pull request by ialisaleh**
This PR introduces the `ArizePhoenixTracer` to enhance LangFlow's observability by integrating with **Arize** and **Phoenix** tracing solutions. The tracer captures spans, logs, and metadata for workflows, providing detailed insights into execution. It supports gRPC and HTTP-based span processing, flow-specific tracing with attributes like `flow_id`, and error handling to ensure smooth operation even if tracing setup fails. The tracer is compatible with LangChain's Instrumentor and serializes inputs, outputs, and logs for seamless integration. This enhancement helps users monitor and debug workflows effectively while maintaining compatibility with existing LangFlow infrastructure.

**Environment Variables**
To enable tracing, users need to configure the following environment variables for Arize and Phoenix:

`ARIZE_PHOENIX_PROJECT` (optional): Specifies the project name for Arize/Phoenix integration.
`ARIZE_PHOENIX_BATCH` (optional): Enables batch processing for span submission.

**Arize Configuration**
`ARIZE_API_KEY`: API Key for Arize integration.
`ARIZE_SPACE_ID`: Space ID for Arize project.
`ARIZE_COLLECTOR_ENDPOINT` (optional): Defaults to [https://otlp.arize.com](https://otlp.arize.com).

**Phoenix Configuration**
`PHOENIX_API_KEY`: API Key for Phoenix integration.
`PHOENIX_COLLECTOR_ENDPOINT` (optional): Defaults to [https://app.phoenix.arize.com](https://app.phoenix.arize.com).

### CODE REVIEW

**Review by italojohnny — APPROVED** [review]

File: src/backend/base/langflow/services/tracing/arize_phoenix.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        self.flow_id = trace_name.split(" - ")[-1]
+
+        try:
+            self._ready = self.setup_arize_phoenix()
+            if not self._ready:
+                return
+
+            self.tracer = self.tracer_provider.get_tracer(__name__)
+            self.propagator = TraceContextTextMapPropagator()
+            self.carrier = {}
```
Comment (italojohnny) [inline src/backend/base/langflow/services/tracing/arize_phoenix.py:53]:
```suggestion
            self.carrier: dict[Any, Any] = {}
```
mypy requires the type to be explicitly defined.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+        if processed_inputs:
+            child_span.set_attribute(SpanAttributes.INPUT_VALUE, self._safe_json_dumps(processed_inputs))
+            child_span.set_attribute(SpanAttributes.INPUT_MIME_TYPE, OpenInferenceMimeTypeValues.JSON.value)
+
+        processed_metadata = self._convert_to_arize_phoenix_types(metadata) if metadata else {}
+        if processed_metadata:
+            for key, value in processed_metadata.items():
+                child_span.set_attribute(f"{SpanAttributes.METADATA}.{key}", value)
+
+        self.child_spans[trace_id] = child_span
```
Comment (italojohnny) [inline src/backend/base/langflow/services/tracing/arize_phoenix.py:204]:
```suggestion
        self.child_spans[UUID(trace_id)] = child_span
```
mypy requires a conversion to match its type definition.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        trace_name: str,
+        outputs: dict[str, Any] | None = None,
+        error: Exception | None = None,
+        logs: Sequence[Log | dict] = (),
+    ) -> None:
+        """Ends a trace span, attaching outputs, errors, and logs as attributes."""
+        if not self._ready or trace_id not in self.child_spans:
+            return
+
+        child_span = self.child_spans[trace_id]
```
Comment (italojohnny) [inline src/backend/base/langflow/services/tracing/arize_phoenix.py:219]:
```suggestion
        child_span = self.child_spans[UUID(trace_id)]
```
mypy requires a conversion to match its type definition.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                    OTELSpanAttributes.EXCEPTION_MESSAGE: exception_message,
+                    OTELSpanAttributes.EXCEPTION_ESCAPED: False,
+                    OTELSpanAttributes.EXCEPTION_STACKTRACE: error_string,
+                }
+                child_span.add_event(name="exception", attributes=attributes)
+        else:
+            child_span.set_status(Status(StatusCode.OK))
+
+        child_span.end(end_time=self._get_current_timestamp())
+        self.child_spans.pop(trace_id)
```
Comment (italojohnny) [inline src/backend/base/langflow/services/tracing/arize_phoenix.py:257]:
```suggestion
        self.child_spans.pop(UUID(trace_id))
```
mypy requires a conversion to match its type definition.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+from typing_extensions import override
+
+from langflow.schema.data import Data
+from langflow.schema.message import Message
+from langflow.services.tracing.base import BaseTracer
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+    from uuid import UUID
```
Comment (italojohnny) [inline src/backend/base/langflow/services/tracing/arize_phoenix.py:26]:
```suggestion
```
Remove this import.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py (hunk 1 above)
Comment (italojohnny) [reply src/backend/base/langflow/services/tracing/arize_phoenix.py:53]: This was just a suggestion. Feel free to define the most appropriate typing, perhaps `dict[Union[str, UUID], Any]`.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
+from __future__ import annotations
+
+import json
+import math
+import os
+import traceback
+import types
+from datetime import datetime, timezone
+from typing import TYPE_CHECKING, Any
```
[the remark this answers is not in the document]
Comment (ialisaleh) [reply src/backend/base/langflow/services/tracing/arize_phoenix.py:9]: We can't move this import from here as it's only used for type-checking.

TCH003 Move standard library import `uuid.UUID` into a type-checking block

File: src/backend/base/langflow/services/tracing/arize_phoenix.py (hunk 3 above)
Comment (ialisaleh) [reply src/backend/base/langflow/services/tracing/arize_phoenix.py:220]: It's not required as I have defined the types for `child_spans` to be `[str, Span]` now.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py (hunk 2 above)
Comment (ialisaleh) [reply src/backend/base/langflow/services/tracing/arize_phoenix.py:205]: It's not required as I have defined the types for `child_spans` to be `[str, Span]` now.

File: src/backend/base/langflow/services/tracing/arize_phoenix.py (hunk 1 above)
Comment (ialisaleh) [reply src/backend/base/langflow/services/tracing/arize_phoenix.py]: I have defined the typing for `carrier` to be `dict[Any, CarrierT]` now.

**Pull request merged**